File: lintchecks/src/main/java/org/signal/lint/AlertDialogBuilderDetector.java
package org.signal.lint;

import com.android.tools.lint.client.api.JavaEvaluator;
import com.android.tools.lint.detector.api.Category;
import com.android.tools.lint.detector.api.Detector;
import com.android.tools.lint.detector.api.Implementation;
import com.android.tools.lint.detector.api.Issue;
import com.android.tools.lint.detector.api.JavaContext;
import com.android.tools.lint.detector.api.LintFix;
import com.android.tools.lint.detector.api.Scope;
import com.android.tools.lint.detector.api.Severity;
import com.intellij.psi.PsiMethod;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.uast.UCallExpression;
import org.jetbrains.uast.UExpression;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

@SuppressWarnings("UnstableApiUsage")
public final class AlertDialogBuilderDetector extends Detector implements Detector.UastScanner {

  static final Issue ALERT_DIALOG_BUILDER_USAGE = Issue.create("AlertDialogBuilderUsage",
                                                               "Creating dialog with AlertDialog.Builder instead of MaterialAlertDialogBuilder",
                                                               "Signal utilizes MaterialAlertDialogBuilder for more consistent and pleasant AlertDialogs.",
                                                               Category.MESSAGES,
                                                               5,
                                                               Severity.WARNING,
                                                               new Implementation(AlertDialogBuilderDetector.class, Scope.JAVA_FILE_SCOPE));

  @Override
  public @Nullable List<String> getApplicableConstructorTypes() {
    return Arrays.asList("android.app.AlertDialog.Builder", "androidx.appcompat.app.AlertDialog.Builder");
  }

  @Override
  public void visitConstructor(JavaContext context, @NotNull UCallExpression call, @NotNull PsiMethod method) {
    JavaEvaluator evaluator = context.getEvaluator();

    if (evaluator.isMemberInClass(method, "android.app.AlertDialog.Builder")) {
      LintFix fix = quickFixIssueAlertDialogBuilder(call);
      context.report(ALERT_DIALOG_BUILDER_USAGE,
                     call,
                     context.getLocation(call),
                     "Using 'android.app.AlertDialog.Builder' instead of com.google.android.material.dialog.MaterialAlertDialogBuilder",
                     fix);
    }

    if (evaluator.isMemberInClass(method, "androidx.appcompat.app.AlertDialog.Builder")) {
      LintFix fix = quickFixIssueAlertDialogBuilder(call);
      context.report(ALERT_DIALOG_BUILDER_USAGE,
                     call,
                     context.getLocation(call),
                     "Using 'androidx.appcompat.app.AlertDialog.Builder' instead of com.google.android.material.dialog.MaterialAlertDialogBuilder",
                     fix);
    }
  }

  private LintFix quickFixIssueAlertDialogBuilder(@NotNull UCallExpression alertBuilderCall) {
    List<UExpression> arguments = alertBuilderCall.getValueArguments();
    UExpression       context   = arguments.get(0);

    String fixSource = "new com.google.android.material.dialog.MaterialAlertDialogBuilder";

    switch (arguments.size()) {
      case 1:
        fixSource += String.format("(%s)", context);
        break;
      case 2:
        UExpression themeOverride = arguments.get(1);
        fixSource += String.format("(%s, %s)", context, themeOverride);
        break;

      default:
        throw new IllegalStateException("MaterialAlertDialogBuilder overloads should have 1 or 2 arguments");
    }

    String               builderCallSource = alertBuilderCall.asSourceString();
    LintFix.GroupBuilder fixGrouper        = fix().group();
    fixGrouper.add(fix().replace().text(builderCallSource).shortenNames().reformat(true).with(fixSource).build());
    return fixGrouper.build();
  }
}

File: lintchecks/src/main/java/org/signal/lint/AlertDialogBuilderDetector.kt
package org.signal.lint

import com.android.tools.lint.detector.api.Category.Companion.MESSAGES
import com.android.tools.lint.detector.api.Detector
import com.android.tools.lint.detector.api.Implementation
import com.android.tools.lint.detector.api.Issue
import com.android.tools.lint.detector.api.JavaContext
import com.android.tools.lint.detector.api.LintFix
import com.android.tools.lint.detector.api.Scope.Companion.JAVA_FILE_SCOPE
import com.android.tools.lint.detector.api.Severity.WARNING
import com.intellij.psi.PsiMethod
import org.jetbrains.uast.UCallExpression

class AlertDialogBuilderDetector : Detector(), Detector.UastScanner {
  override fun getApplicableConstructorTypes(): List<String> {
    return listOf("android.app.AlertDialog.Builder", "androidx.appcompat.app.AlertDialog.Builder")
  }

  override fun visitConstructor(context: JavaContext, node: UCallExpression, constructor: PsiMethod) {
    val evaluator = context.evaluator

    if (evaluator.isMemberInClass(constructor, "android.app.AlertDialog.Builder")) {
      context.report(
        issue = ALERT_DIALOG_BUILDER_USAGE,
        scope = node,
        location = context.getLocation(node),
        message = "Using 'android.app.AlertDialog.Builder' instead of com.google.android.material.dialog.MaterialAlertDialogBuilder",
        quickfixData = quickFixIssueAlertDialogBuilder(node)
      )
    }

    if (evaluator.isMemberInClass(constructor, "androidx.appcompat.app.AlertDialog.Builder")) {
      context.report(
        issue = ALERT_DIALOG_BUILDER_USAGE,
        scope = node,
        location = context.getLocation(node),
        message = "Using 'androidx.appcompat.app.AlertDialog.Builder' instead of com.google.android.material.dialog.MaterialAlertDialogBuilder",
        quickfixData = quickFixIssueAlertDialogBuilder(node)
      )
    }
  }

  private fun quickFixIssueAlertDialogBuilder(alertBuilderCall: UCallExpression): LintFix {
    val arguments = alertBuilderCall.valueArguments
    val context = arguments[0]

    var fixSource = "new com.google.android.material.dialog.MaterialAlertDialogBuilder"

    when (arguments.size) {
      1 -> fixSource += String.format("(%s)", context)
      2 -> {
        val themeOverride = arguments[1]
        fixSource += String.format("(%s, %s)", context, themeOverride)
      }

      else -> throw IllegalStateException("MaterialAlertDialogBuilder overloads should have 1 or 2 arguments")
    }

    return fix()
      .group()
      .add(
        fix()
          .replace()
          .text(alertBuilderCall.asSourceString())
          .shortenNames()
          .reformat(true)
          .with(fixSource)
          .build()
      )
      .build()
  }

  companion object {
    val ALERT_DIALOG_BUILDER_USAGE: Issue = Issue.create(
      id = "AlertDialogBuilderUsage",
      briefDescription = "Creating dialog with AlertDialog.Builder instead of MaterialAlertDialogBuilder",
      explanation = "Signal utilizes MaterialAlertDialogBuilder for more consistent and pleasant AlertDialogs.",
      category = MESSAGES,
      priority = 5,
      severity = WARNING,
      implementation = Implementation(AlertDialogBuilderDetector::class.java, JAVA_FILE_SCOPE)
    )
  }
}